pkg/core/releaser.go

```diff
@@ -0,0 +1,450 @@
+	var wg sync.WaitGroup
+	errors := make(chan rolloutError)
+	done := make(chan bool)
+	for _, workload := range appliedWorkloads {
+		wg.Add(1)
+		go r.goWatch(workload, errors, &wg)
+	}
+	go goWait(&wg, done)
+	select {
+	case <-done:
+		return nil
+	case err := <-errors:
+		scope, logger := err.resource.scopes(r)
+		return ErrorContext{err: err.err, scope: scope, logger: logger, context: "watch workload"}
+	}
+}
+
+func (r releaser) watchRollback(appliedWorkloads []appResource) []error {
+	var wg sync.WaitGroup
+	errorChan := make(chan rolloutError)
+	done := make(chan bool)
+	for _, workload := range appliedWorkloads {
+		wg.Add(1)
+		go r.goWatch(workload, errorChan, &wg)
+	}
+	var errors []error
+
+	go goWait(&wg, done)
+	for range appliedWorkloads {
+		select {
+		case <-done:
+			return errors
+		case err := <-errorChan:
+			scope, logger := err.resource.scopes(r)
+			errors = append(errors, ErrorContext{err: err.err, scope: scope, logger: logger, context: "watch rollback"})
+		}
+	}
+
+	return errors
+}
+
+func goWait(wg *sync.WaitGroup, done chan bool) {
+	wg.Wait()
+	done <- true
+}
+
+// TODO: add support for watching pods
+// TODO: add support for watching argo rollouts
+func (r releaser) goWatch(resource appResource, errors chan rolloutError, wg *sync.WaitGroup) {
+	if resource.supportsRollback() && !resource.isRollout() {
+		err := k8s.RolloutStatus(resource.kind, resource.name, r.app.Name)
+		if err != nil {
+			errors <- rolloutError{err: err, resource: resource}
+		}
+	}
+	wg.Done()
+}
+
+func (r releaser) rollback(appliedResources []appResource, cachedResources []appResource) ([]appResource, []error) {
+	var rolledBack []appResource
+	var errors []error
+	for _, applied := range appliedResources {
+		var inPreviousState bool
+		scope, logger := applied.scopes(r)
+
+		for _, cached := range cachedResources {
+			if applied.kind == cached.kind && applied.name == cached.name {
+				inPreviousState = true
+				err := r.rollbackResource(applied, cached)
+				if err != nil {
+					errors = append(errors, r.releaseError(ErrorContext{err: err, context: "rollback", scope: scope, logger: logger}))
+					break
+				}
+				rolledBack = append(rolledBack, applied)
+				break
+			}
+		}
+		if !inPreviousState {
+			err := k8s.Delete(applied.kind, applied.name, r.app.Name)
+			if err != nil {
+				errors = append(errors, r.releaseError(ErrorContext{err: err, context: "deleting newly created resource on error", scope: scope, logger: logger}))
+			}
+		}
+	}
+	return rolledBack, errors
+}
+
+func (r releaser) rollbackResource(applied appResource, cached appResource) error {
+	var err error
+	if applied.supportsRollback() {
+		if applied.isRollout() {
+			// TODO: add actual argo support
+			err = k8s.Apply(cached.contents, r.app.Name)
+		} else {
+			err = k8s.RolloutUndo(applied.kind, applied.name, r.app.Name)
+		}
+	} else {
+		err = k8s.Apply(cached.contents, r.app.Name)
+	}
```

Comment:
Is this written this way for your TODO around argo support? If not we could simplify?

```suggestion
	if applied.supportsRollback() && !applied.isRollout() {
			err = k8s.RolloutUndo(applied.kind, applied.name, r.app.Name)
	} else {
		err = k8s.Apply(cached.contents, r.app.Name)
	}
```

Reply:
totally for the TODO - which will be resolved before this exits draft status if all goes according to plan